File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = [0] * n
        prefix = [1] + prefix
        for i in range(1, n + 1):
            prefix[i] = prefix[i - 1] * nums[i - 1]

        postfix = [0] * n
        postfix += [1]
        for i in range(n - 1, -1, -1):
            postfix[i] = postfix[i + 1] * nums[i]
        prefix.pop(0)
        postfix.pop()
        res = []
        for i in range(n):
            left = 1 if i == 0 else prefix[i - 1]
            right = 1 if i == n - 1 else postfix[i + 1]

            res.append(left * right)
        return res

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] prefix = new int[n];
        int[] postfix = new int[n];
        int[] res = new int[n];

        prefix[0] = 1;
        postfix[n - 1] = 1;

        for (int i=1; i<n; i++) {
            prefix[i] = nums[i - 1] * prefix[i - 1];
        }

        for (int j=n-2; j>=0; j--) {
            postfix[j] = nums[j + 1] * postfix[j + 1];
        }

        for (int k=0; k<n; k++) {
            res[k] = prefix[k] * postfix[k];
        }
        return res;
    }
}